libcob/move.c

```diff
@@ -1084,14 +1084,26 @@ optimized_move_display_to_edited (cob_field *f1, cob_field *f2)
 			}
 	}
 
-	/* first check for BLANK WHEN ZERO attribute */
+	/* first check for BLANK WHEN ZERO attribute	*/
+	/* Note that if the src field is signed then we	*/
+	/* scan for one less byte			*/
 	if (COB_FIELD_BLANK_ZERO (f2)) {
-		for (; (src <= src_end) ; src++) {
-			if (*src != '0') break;
-		}
-		if (src > src_end) {
-			memset (dst, (int)' ', (size_t)f2->size);
-			return;
+		if (COB_FIELD_HAVE_SIGN (f1) && COB_FIELD_SIGN_SEPARATE (f1)) {
+			for (; (src < src_end) ; src++) {
+				if (*src != '0') break;
+			}
+			if (src >= src_end) {
+				memset (dst, (int)' ', (size_t)f2->size);
+				return;
+			}
+		} else {
+			for (; (src <= src_end) ; src++) {
+				if (*src != '0') break;
+			}
+			if (src > src_end) {
+				memset (dst, (int)' ', (size_t)f2->size);
+				return;
+			}
```

Comment:
Please change it to use something like `const unsigned char *check_end = !COB_FIELD_SIGN_SEPARATE (f1) ? src_end : src_end - 1;` instead to reduce the duplication.

Also: What about separate and leading sign?
Note: I _think_ there's no use in explicit casting a char to int, is it? The field->size is size_t already, so that should be dropped in any case.

Reply:
> Also: What about separate and leading sign? And... where are the overpunched positions checked?

There is an invariant here that `f1` either has no sign or has a trailing separate sign (as checked by the block between `#ifndef NDEBUG` block at the top of the function, and also mentionned by Chuck in his message).

> As we just had that in the other PR: What about invalid data - should we check with `COB_D2I() = 0` instead?

We could add that, though the other PR is about moving from an ALPHANUMERIC source, while `optimized_move_display_to_edited` only moves from a NUMERIC-DISPLAY source (either a direct one, or a temporary one).
 
> Note: I _think_ there's no use in explicit casting a char to int, is it?

Absolutely, those character literals should already be of type `int` anyways.

> The field->size is size_t already, so that should be dropped in any case.

Done.

Reply:
> Can you expand that to also the leading separate case (_should_ be simple, and we now even have a testcase ready :-) and to the overpunched sign (a bit more complex)? Of course we can postpone that to later (either both or one of those)...

Done for this part.

> Just checked with MF: After `MOVE SPACES TO L(1:3)` before the W-X move, by default we get the abort for non-numeric data in numeric displays, when I disable that at run-time, then " 00000" is blanked to space and " 1234" results in " 12,34" while an additional `MOVE "A" TO L (8:1)` results in
> 
> ```
>  ===>       0,01<===
>  ===>      12,31<===
> ```

Added this to the test case, though the result differs, i.e.:
```
           MOVE 0 TO L.
           MOVE SPACES TO L(1:3).
           MOVE L TO W-X.
           DISPLAY ' ===>' W-X '<==='.

           MOVE "A" TO L(8:1).
           DISPLAY ' ===>' W-X '<==='.

           MOVE 1234 TO L.
           MOVE SPACES TO L(1:3).
           MOVE L TO W-X.
           DISPLAY ' ===>' W-X '<==='.

           MOVE "A" TO L(8:1).
           DISPLAY ' ===>' W-X '<==='.
```

Yields:
```
 ===>           <===
 ===>           <===
 ===>     012,34<===
 ===>     012,34<===
```

Though this is not really related to the "BLANK WHEN ZERO not working anymore" issue.